src/libraries/Common/src/System/Net/Http/Http3/Helpers/VariableLengthIntegerHelper.cs

```diff
@@ -0,0 +1,213 @@
+                            Debug.Assert(value >= 0 && value <= EightByteLimit, "Serialized values are within [0, 2^62).");
+
+                            bytesRead = 8;
+                            return true;
+                        }
+                        break;
+                }
+            }
+
+            value = 0;
+            bytesRead = 0;
+            return false;
+        }
+
+        public static bool TryRead(ref SequenceReader<byte> reader, out long value)
+        {
+            // Hot path: we probably have the entire integer in one unbroken span.
+            if (TryRead(reader.UnreadSpan, out value, out int bytesRead))
+            {
+                reader.Advance(bytesRead);
+                return true;
+            }
+
+            // Cold path: copy to a temporary buffer before calling span-based read.
+            return TryReadSlow(ref reader, out value);
+
+            static bool TryReadSlow(ref SequenceReader<byte> reader, out long value)
+            {
+                ReadOnlySpan<byte> span = reader.CurrentSpan;
+
+                if (reader.TryPeek(out byte firstByte))
+                {
+                    int length =
+                        (firstByte & LengthMask) switch
+                        {
+                            LengthOneByte => 1,
+                            LengthTwoByte => 2,
+                            LengthFourByte => 4,
+                            _ => 8 // LengthEightByte
+                    };
+
+                    Span<byte> temp = (stackalloc byte[8])[..length];
+                    if (reader.TryCopyTo(temp))
+                    {
+                        bool result = TryRead(temp, out value, out int bytesRead);
+                        Debug.Assert(result == true);
+                        Debug.Assert(bytesRead == length);
+
+                        reader.Advance(bytesRead);
+                        return true;
+                    }
+                }
+
+                value = 0;
+                return false;
+            }
+        }
+
+        public static long GetInteger(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+        {
+            var reader = new SequenceReader<byte>(buffer);
+            if (TryRead(ref reader, out long value))
+            {
+                consumed = examined = buffer.GetPosition(reader.Consumed);
+                return (long)value;
+            }
+            else
+            {
+                consumed = default;
+                examined = buffer.End;
+                return -1;
+            }
+        }
+
+        public static bool TryWrite(Span<byte> buffer, long longToEncode, out int bytesWritten)
+        {
+            Debug.Assert(longToEncode >= 0);
+            Debug.Assert(longToEncode <= EightByteLimit);
+
+            if (longToEncode < OneByteLimit)
+            {
+                if (!buffer.IsEmpty)
```

Comment:
Nit: in other places like https://github.com/dotnet/runtime/pull/1294/files#diff-fb1bdc75c84ead7a94a9c379a52f6230R46, we've used `buffer.Length != 0`.  It'd be nice to be consistent one way or the other.

Reply:
@gfoidl IIRC you mentioned that checking `buffer.Length != 0` is faster than `buffer.IsEmpty`. Is that correct?

Reply:
It results in the same codegen -- so no perf-difference. My suggestion was about readability / expressiveness, and IMHO `buffer.IsEmpty` does a good job therefore.